### PR TITLE
python3Packages.nanobind: init at 1.9.0

### PR DESCRIPTION
**Pull request by phiadaarr**
## Description of changes

nanobind is a small binding library that exposes C++ types in Python and vice versa. It is reminiscent of [Boost.Python](https://www.boost.org/doc/libs/1_64_0/libs/python/doc/html) and [pybind11](https://github.com/pybind/pybind11) and uses near-identical syntax. In contrast to these existing tools, nanobind is more efficient: bindings compile in a shorter amount of time, produce smaller binaries, and have better runtime performance.

https://github.com/wjakob/nanobind

Pinging the pybind11 maintainers @yuriaisaka @dotlambda
Pinging the nanobind/pybind11 author @wjakob 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux (except tests depending on torch since cache.nixos.org does not provide a binary for torch and I did not have the patience to build it)
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/nanobind/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/nanobind/default.nix:8]:
```suggestion
  pyproject = true;
```

File: pkgs/development/python-modules/nanobind/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/nanobind/default.nix:45]:
```suggestion
    description = "Tiny and efficient C++/Python bindings";
```

Comment (phiadaarr) [reply pkgs/development/python-modules/nanobind/default.nix]: @dotlambda thanks for you incredibly quick review - highly appreciated

File: pkgs/development/python-modules/nanobind/default.nix
Comment (phiadaarr) [reply pkgs/development/python-modules/nanobind/default.nix]: Done

File: pkgs/development/python-modules/nanobind/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/nanobind/default.nix:2]:
```suggestion
{ lib, buildPythonPackage, pythonOlder, fetchFromGitHub, cmake, eigen
, ninja, scikit-build, pytestCheckHook, numpy, scipy, torch, jax, jaxlib
```

**Comment by phiadaarr** [comment]
Is there anything, I can do on my side such that we can merge this, @dotlambda ? Many thanks for your review!

**Comment by dotlambda** [comment]
Please ping tensorflow maintainers to fix the remaining error.

File: pkgs/development/python-modules/nanobind/default.nix
Comment (stephen-huan) [inline pkgs/development/python-modules/nanobind/default.nix:14]: @phiadaarr do you know why the submodule vendored dependency can't be removed with `-DNB_USE_SUBMODULE_DEPS=OFF` and adding `pkgs.robin-map` to `buildInputs` like so?

```nix
  pname = "nanobind";
  version = "1.9.0";
  pyproject = true;

  src = fetchFromGitHub {
    owner = "wjakob";
    repo = pname;
    rev = "v${version}";
    hash = "sha256-NB14GMWKTMgxMLpMwF44fNW0fyjACSJxPhEp/xKYvnE=";
    fetchSubmodules = false;
  };

  disabled = pythonOlder "3.8";

  nativeBuildInputs = [ cmake ninja scikit-build setuptools ];
  buildInputs = [ eigen robin-map ];
  dontUseCmakeBuildDir = true;
  cmakeFlags = [
    "-DNB_USE_SUBMODULE_DEPS=OFF"
  ];
```

when I try compiling [pypcode](https://discourse.nixos.org/t/cmake-3-27-8-nix-support-setup-hook-line-126-cmake-command-not-found/39053/2) against the above it gives me header warnings such as

```text
In file included from /nix/store/h52r8rx5rdxak3ij9v5zz6rh8gvw2phv-python3.11-nanobind-1.9.0/lib/python3.11/site-packages/nanobind/src/nb_func.cpp:10:
/nix/store/h52r8rx5rdxak3ij9v5zz6rh8gvw2phv-python3.11-nanobind-1.9.0/lib/python3.11/site-packages/nanobind/src/nb_internals.h:14:10: fatal error: nanobind/nanobind.h: No such file or directory
   14 | #include <nanobind/nanobind.h>
      |          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

Instead, I have to place `robin-map` where it expects the submodule.

```nix
  pname = "nanobind";
  version = "1.9.0";
  pyproject = true;

  src = fetchFromGitHub {
    owner = "wjakob";
    repo = pname;
    rev = "v${version}";
    hash = "sha256-NB14GMWKTMgxMLpMwF44fNW0fyjACSJxPhEp/xKYvnE=";
    fetchSubmodules = false;
  };

  disabled = pythonOlder "3.8";

  nativeBuildInputs = [ cmake ninja scikit-build setuptools ];
  buildInputs = [ eigen robin-map ];
  dontUseCmakeBuildDir = true;
  preConfigure = ''
    mkdir -p ext/robin_map
    cp -r ${robin-map}/include -t ext/robin_map
  '';
```

File: pkgs/development/python-modules/nanobind/default.nix
Comment (stephen-huan) [inline pkgs/development/python-modules/nanobind/default.nix:6]: Version [1.9.0](https://github.com/wjakob/nanobind/releases/tag/v1.9.0) is out and the derivation seems to work if only the hash is changed.

**Review by stephen-huan — CHANGES_REQUESTED** [review]

**Comment by phiadaarr** [comment]
@abbradar pinging you in your capacity as the maintainer of the tensorflow package. When I try to create an environment in which both `torch` and `tensorflow` are present, I get:

```
Running phase: pythonCatchConflictsPhase
Found duplicated packages in closure for dependency 'tensorboard': 
        tensorboard 2.15.1 (/nix/store/14gikl62hxwdqdwhfndahr8x71arbj92-python3.11-tensorboard-2.15.1/lib/python3.11/site-packages/tensorboard-2.15.1.dist-info)
        tensorboard 2.15.1 (/nix/store/v9yv7sgvkgmm0dgdgl038bdsk43cggjy-python3.11-tensorboard-2.15.1/lib/python3.11/site-packages/tensorboard-2.15.1.dist-info)
Found duplicated packages in closure for dependency 'protobuf': 
        protobuf 4.24.4 (/nix/store/m61dvbg0ncrf08mghn4n3r9x4cyq7dgq-python3.11-protobuf-4.24.4/lib/python3.11/site-packages/protobuf-4.24.4.dist-info)
        protobuf 4.21.12 (/nix/store/vkdvmsjz81x2xlws8qnkw4dqg51d2l8r-python3.11-protobuf-4.21.12/lib/python3.11/site-packages/protobuf-4.21.12.dist-info)
Found duplicated packages in closure for dependency 'grpcio': 
        grpcio 1.60.1 (/nix/store/ccx0r2jzx5wxsp2dqa9a35l8amlspdvl-python3.11-grpcio-1.60.1/lib/python3.11/site-packages/grpcio-1.60.1.dist-info)
        grpcio 1.60.1 (/nix/store/asq5z2xncj4wp8k10igqlc99a2smmyax-python3.11-grpcio-1.60.1/lib/python3.11/site-packages/grpcio-1.60.1.dist-info)
Found duplicated packages in closure for dependency 'tensorboard_plugin_profile': 
        tensorboard_plugin_profile 2.11.1 (/nix/store/b84ndcxxald19yj2yy1sqbzd68cdvfg5-python3.11-tensorboard_plugin_profile-2.11.1/lib/python3.11/site-packages/tensorboard_plugin_profile-2.11.1.dist-info)
        tensorboard_plugin_profile 2.11.1 (/nix/store/m5m47mw6gk7n55k8yz709n7r4ybdy1rd-python3.11-tensorboard_plugin_profile-2.11.1/lib/python3.11/site-packages/tensorboard_plugin_profile-2.11.1.dist-info)

Package duplicates found in closure, see above. Usually this happens if two packages depend on different version of the same dependency.
```

One possibility would be, I believe, to disable tensorboard support in `tensorflow`. However, this causes a massive rebuild that needs more RAM than my machine provides.

What is your take on this? Is this a common problem? Would it be possible to solve it? Or would be disabling tensorboard support the way to go?

Many thanks in advance - your help is highly appreciated.

**Comment by phiadaarr** [comment]
@abbradar just a quick reminder. Do you have any idea how to solve this duplicate dependency issue? Otherwise, I would propose to disable the tests (only a tiny fraction of the overall test suite) and merge nanobind without them. Many thanks for your time!

**Comment by phiadaarr** [comment]
@dotlambda I believe we can merge now. The tensorflow dependency issues have been resolved on `master`. Awesome!

File: pkgs/development/python-modules/nanobind/default.nix
Comment (phiadaarr) [reply pkgs/development/python-modules/nanobind/default.nix:14]: Sorry for the late response. As you have correctly figured out, the build system of nanobind expects `robin_map` to be located within `ext/`. If you put it somewhere else, the build will fails.

Comment (stephen-huan) [reply pkgs/development/python-modules/nanobind/default.nix:14]: @phiadaarr  I'm trying to avoid the vendored `robin_map` dependency pulled in with `fetchSubmodules = true`. In [`CMakeLists.txt`](https://github.com/wjakob/nanobind/blob/55bc50a797b5578d724701c97a8b647ca3b6f2bc/CMakeLists.txt#L15) there's a configuration option `NB_USE_SUBMODULE_DEPS` which controls whether nanobind should "Use the nanobind dependencies shipped as a git submodule of this repository". If this option is turned off, then [`cmake/nanobind-config.cmake`](https://github.com/wjakob/nanobind/blob/55bc50a797b5578d724701c97a8b647ca3b6f2bc/cmake/nanobind-config.cmake#L206-L219) searches for `robin_map` with `find_dependency(tsl-robin-map)`.

```cmake
# If nanobind was installed without submodule dependencies, then the
# dependencies directory won't exist and we need to find them.
# However, if the directory _does_ exist, then the user is free to choose
# whether nanobind uses them (based on `NB_USE_SUBMODULE_DEPS`), with a
# preference to choose them if `NB_USE_SUBMODULE_DEPS` is not defined
if (NOT IS_DIRECTORY ${NB_DIR}/ext/robin_map/include OR
    (DEFINED NB_USE_SUBMODULE_DEPS AND NOT NB_USE_SUBMODULE_DEPS))
  include(CMakeFindDependencyMacro)
  find_dependency(tsl-robin-map)
  target_link_libraries(${TARGET_NAME} PRIVATE tsl::robin_map)
else()
  target_include_directories(${TARGET_NAME} PRIVATE
    ${NB_DIR}/ext/robin_map/include)
endif()
```

That's why I'm confused why just adding `robin-map` to `buildInputs` and `"-DNB_USE_SUBMODULE_DEPS=OFF"` to `cmakeFlags` doesn't just work. The compilation succeeds but when I try to build [pypcode](https://discourse.nixos.org/t/cmake-3-27-8-nix-support-setup-hook-line-126-cmake-command-not-found/39053/2) e.g. against nanobind I get header not found errors. I'm not sure how one is supposed to feed dependencies to CMake otherwise.

Of course this discussion is mostly theoretical as one can just put `robin_map` where nanobind expects the submodule to avoid the CMake machinery which works and also avoids the vendored dependency.

```nix
  preConfigure = ''
    mkdir -p ext/robin_map
    cp -r ${robin-map}/include -t ext/robin_map
  '';
```

I asked the question in case there was something simple I was missing as I'm new to CMake. But concretely, with respect to the PR, is there a reason you didn't want to avoid the vendored dependency with the above snippet?

Comment (wjakob) [reply pkgs/development/python-modules/nanobind/default.nix:14]: Hi @phiadaarr ,

👋 Author here  ;-). Indeed, it would be good if you could leave this untouched since it affects ABI stability. If you have a Python process that imports multiple nanobind-based packages, the nanobind lib in each extension will talk to the others, and they will exchange internal data structures to enable inter-module data structure exchange. The robin_map data structure is part of this ABI contract. I am not saying that this is likely, but two identical nanobind versions referencing incompatible robin_map versions could hypothetically lead to undefined behavior.

Comment (wjakob) [reply pkgs/development/python-modules/nanobind/default.nix:14]: Thanks for making a nix package by the way!

Comment (stephen-huan) [reply pkgs/development/python-modules/nanobind/default.nix:14]: > Since the author of nanobind chose to pin an exact version of robin_map via a submodule, I believe we should follow this for nixpkgs and check out the exact version the author wants us to use.

This is subjective, but my impression of the philosophy of nixpkgs was to remove vendored dependencies wherever possible---as nixpkgs is a means of dependency specification itself, to keep software up to date, and to avoid the maintenance burden of keeping multiple versions around. For example, every python package ignores the canonical ways of specifying dependencies (`requirements.txt`, `pyproject.toml`, etc.) by necessity since multiple versions of the same python package conflict. From a user perspective, this decision makes tracking and manipulating dependencies (e.g. overriding, overlaying) transparent. Of course, the trade-off is that updates to dependencies can break parents, but generally nixpkgs skews heavily in favor of keeping software up to date.

On more practical note, the [pinned version](https://github.com/Tessil/robin-map/tree/68ff7325b3898fca267a103bad5c509e8861144d) is exactly the [latest version](https://github.com/Tessil/robin-map/releases/tag/v1.2.1) on nixpkgs and future breakage seems unlikely. However, I also don't see a compelling reason to overlay robin-map, so I agree that this is not a blocker.

> Sorry that I am not of much help here.

No worries! Apologies myself for the noise :p

Comment (wjakob) [reply pkgs/development/python-modules/nanobind/default.nix:14]: @stephen-huan did you see the example situation where a vendored dependency could potentially lead to ABI breakage and crashes in combination with other packages that haven't been built with nix?

Comment (stephen-huan) [reply pkgs/development/python-modules/nanobind/default.nix:14]: I wrote my comment before I saw yours, sorry. Your reasoning makes sense.

Comment (phiadaarr) [reply pkgs/development/python-modules/nanobind/default.nix:14]: Thanks for jumping in @wjakob! I think we can consider this resolved.

**Comment by phiadaarr** [comment]
I think we can merge now. Pinging you, @NickCao, as you have previously merged similar contributions by me.

File: pkgs/development/python-modules/nanobind/default.nix
Comment (NickCao) [inline pkgs/development/python-modules/nanobind/default.nix:1]: Please format with `nixfmt-rfc-style`

File: pkgs/development/python-modules/nanobind/default.nix
Comment (NickCao) [inline pkgs/development/python-modules/nanobind/default.nix:24]: Or should it be `preCheck`? Since it's only for the tests.

Comment (phiadaarr) [reply pkgs/development/python-modules/nanobind/default.nix:44]: Indeed, thanks for the comment.

File: pkgs/development/python-modules/nanobind/default.nix
Comment (phiadaarr) [reply pkgs/development/python-modules/nanobind/default.nix]: Done

**Comment by NickCao** [comment]
@ofborg build python311Packages.nanobind python312Packages.nanobind

**Comment by phiadaarr** [comment]
Is there anything I can do such that we can merge this? Any pointers highly appreciated :)

**Comment by kirillrdy** [comment]
greetings @phiadaarr  !

I am interested in getting this merged, unfortunately currently one of dependecis fails to build on master

`pythonPackages.protobuf`

I certainly don't expect/require you to fix this for this PR, but any help would be appreciated

**Comment by kirillrdy** [comment]
Result of `nixpkgs-review pr 286125` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python312Packages.nanobind</li>
    <li>python312Packages.nanobind.dist</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python311Packages.nanobind</li>
    <li>python311Packages.nanobind.dist</li>
  </ul>
</details>

protobuf build log

```
error: builder for '/nix/store/xp69yxchzp0hdy9jgfl0agmz9nq2hsqz-python3.12-protobuf-4.21.12.drv' failed with exit code 1;
       last 10 log lines:
       > google/protobuf/internal/well_known_types_test.py::TimeUtilTest::testTimestampSerializeAndParse
       >   /build/source/python/google/protobuf/internal/well_known_types.py:195: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
       >     self.FromDatetime(datetime.datetime.utcnow())
       >
       > -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
       > =========================== short test summary info ============================
       > FAILED google/protobuf/internal/json_format_test.py::JsonFormatTest::testInvalidOneof - AttributeError: 'JsonFormatTest' object has no attribute 'assertRaisesRegex...
       > FAILED google/protobuf/internal/json_format_test.py::JsonFormatTest::testInvalidTimestamp - AttributeError: 'JsonFormatTest' object has no attribute 'assertRaisesRegex...
       > ============= 2 failed, 900 passed, 9 skipped, 9 warnings in 7.09s =============
       > /nix/store/v5lsd029lz5lfhamivbgqyp3zdv94ah2-stdenv-linux/setup: line 1578: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/xp69yxchzp0hdy9jgfl0agmz9nq2hsqz-python3.12-protobuf-4.21.12.drv'.

```

File: pkgs/development/python-modules/nanobind/default.nix
Comment (kirillrdy) [inline pkgs/development/python-modules/nanobind/default.nix:26]: use of pname is considered an anti-pattern https://github.com/NixOS/nixpkgs/issues/277994

Comment (phiadaarr) [reply pkgs/development/python-modules/nanobind/default.nix:26]: From skimming through the issue, it is not obvious to me what the problem is and how the fix would look like for nanobind. Do you have any insights on this?

Comment (kirillrdy) [reply pkgs/development/python-modules/nanobind/default.nix:26]: basically it's about usage of pname ,when it's not really linked.

eg in `repo = pname,` if someone forks a packages and keeps name

**Comment by phiadaarr** [comment]
Regarding the `protobuf` build failing:

- I tried `nix-build -A python3Packages.protobuf --check` on the latest master (c077b043ee73cb0c479c7c8d1d01e42c71a8372f) and there `protobuf` builds and the tests run through.
- I ran `nix-build -A python3Packages.nanobind` on my nanobind branch -> works

Not sure what the problem overall is and whether it is a problem after all for us or not.

Is there anything that needs to be fixed from our side before we can merge this package, @NickCao @dotlambda ?

**Comment by kirillrdy** [comment]
Result of `nixpkgs-review pr 286125` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python312Packages.nanobind</li>
    <li>python312Packages.nanobind.dist</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python311Packages.nanobind</li>
    <li>python311Packages.nanobind.dist</li>
  </ul>
</details>


but now python312Packages.nanobind fails because python312Packages.tensorflow fails to build

I think, this should be good to merge

**Review by kirillrdy — APPROVED** [review]

**Comment by phiadaarr** [comment]
Is there anything, I can do to get this merged? Any pointers highly appreciated :)

**Comment by kirillrdy** [comment]
@GrahamcOfBorg build python3Packages.nanobind

**Comment by kirillrdy** [comment]
@phiadaarr thank you for this !

**Pull request merged**
